packages/teuchos/parameterlist/src/Teuchos_ParameterListModifier.cpp

```diff
@@ -0,0 +1,213 @@
+}
+
+
+Array<std::string> ParameterListModifier::findMatchingBaseNames(const ParameterList &pl,
+    const std::string &base_name, const bool &find_parameters, const bool &find_sublists) const{
+  Array<std::string> matches(0);
+  ParameterList::ConstIterator itr;
+  for (itr = pl.begin(); itr != pl.end(); ++itr) {
+    const std::string &name = pl.name(itr);
+    std::size_t found = name.find(base_name);
+    if (found == 0){
+      if (pl.isSublist(name)){
+        if (find_sublists){
+          matches.push_back(name);
+        }
+      } else{
+        if (find_parameters){
+          matches.push_back(name);
+        }
+      }
+    }
+  }
+  return matches;
+}
+
+
+int ParameterListModifier::expandParameters(
+    const std::string &param_template_name, ParameterList &pl, ParameterList &valid_pl,
+    const Array<std::string> &exclude_parameters) const
+{
+  int replacements = 0;
+  auto ignore_names = exclude_parameters;
+  std::sort(ignore_names.begin(), ignore_names.end());
+  ParameterList::ConstIterator itr;
+  if (valid_pl.isParameter(param_template_name)){
+    ParameterEntry &valid_pl_entry = valid_pl.getEntry(param_template_name);
+    for (itr = pl.begin(); itr != pl.end(); ++itr) {
+      const std::string &param_name = pl.name(itr);
+      if (!pl.isSublist(param_name)){
+        if (!std::binary_search(ignore_names.begin(), ignore_names.end(), param_name)){
+          valid_pl.setEntry(param_name, valid_pl_entry);
+          replacements += 1;
+        }
+      }
+    }
+    valid_pl.remove(param_template_name);
+  }
+  return replacements;
+}
+
+
+int ParameterListModifier::expandSublists(
+    const std::string &sublist_template_name, ParameterList &pl, ParameterList &valid_pl,
+    const Array<std::string> &exclude_sublists) const
+{
+  int replacements = 0;
+  auto ignore_names = exclude_sublists;
+  std::sort(ignore_names.begin(), ignore_names.end());
+  ParameterList::ConstIterator itr;
+  if (valid_pl.isSublist(sublist_template_name)){
+    ParameterList &valid_pl_sublist = valid_pl.get<ParameterList>(sublist_template_name);
+    for (itr = pl.begin(); itr != pl.end(); ++itr) {
+      const std::string &subname = pl.name(itr);
+      if (pl.isSublist(subname)){
+        if (!std::binary_search(ignore_names.begin(), ignore_names.end(), subname)){
+          valid_pl.set(subname, valid_pl_sublist);
+          replacements += 1;
+        }
+      }
+    }
+    valid_pl.remove(sublist_template_name);
+  }
+  return replacements;
+}
+
+
+int ParameterListModifier::expandSublistsUsingBaseName(
+    const std::string &base_name, ParameterList &pl, ParameterList &valid_pl,
+    const bool &allow_base_name) const
+{
+  int replacements = 0;
+  bool delete_base_name = true;
+  if (valid_pl.isSublist(base_name)){
+    if (pl.isSublist(base_name)){
+      if (!allow_base_name){
+        throw std::logic_error("Sublist can't have the same name as the parameter template name without `allow_base_name=true`.");
```

Comment:
Please use `TEUCHOS_TEST_FOR_EXCEPTION()` instead of raw `throw`.

Reply:
Fixed in next push